onnxruntime/contrib_ops/cpu/optional_ops/optional_ops.cc

```diff
@@ -0,0 +1,175 @@
+    CopyCpuTensor(&*in_tensor, &tmp);
+    output_tensors.push_back(std::move(tmp));
+  }
+
+  auto ml_tensor_sequence = DataTypeImpl::GetType<TensorSeq>();
+  clone.Init(output_sequence_tensors.release(),
+             ml_tensor_sequence,
+             ml_tensor_sequence->GetDeleteFunc());
+
+  return;
+}
+
+static bool CheckValidTypeProto(const onnx::TypeProto& tp) {
+  // Optional types can currently be Tensors or SequenceTensors only
+  // Ensure that the TypeProto holds those types
+
+  // TODO: Should we also ensure that element type for tensor is set ?
+  return (tp.has_tensor_type()) ||
+         (tp.has_sequence_type() && tp.sequence_type().elem_type().has_tensor_type());
+}
+
+Status Optional::Compute(OpKernelContext* ctx) const {
+  const auto* input_ort_value = ctx->GetInputOrtValue(0);
+  auto* output_ort_value = ctx->GetOutputOrtValue(0);
+
+  // The static allocation planner has deemed that the input can be re-used as the output
+  // Analogy: Checking if data pointers for the input and output Tensors are the same
+  // before proceeding to make the copy.
+  if (output_ort_value->IsAllocated()) {
+    return Status::OK();
+  }
+
+  if (input_ort_value != nullptr) {  // An input was provided by the user
+
+    AllocatorPtr alloc;
+    ORT_RETURN_IF_ERROR(ctx->GetTempSpaceAllocator(&alloc));
+
+    if (input_ort_value->IsTensor()) {
+      CloneTensor(alloc,
+                  input_ort_value->Get<Tensor>(), *output_ort_value);
+    } else if (input_ort_value->IsTensorSequence()) {
+      CloneSequnceTensor(alloc,
+                         input_ort_value->Get<TensorSeq>(), *output_ort_value);
+    } else {
+      // Will not reach here
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Only Tensors and Sequence Tensors can be used to construct an Optional type");
+    }
+
+  } else {  // No input was provided - we use the type proto to construct the output OrtValue
+
+    CheckValidTypeProto(*type_proto_);
+
+    // type is either Tensor or TensorSeq (we have validated this already in CheckValidTypeProto())
+    auto type = type_proto_->has_tensor_type() ? DataTypeImpl::GetType<Tensor>() : DataTypeImpl::GetType<TensorSeq>();
+
+    output_ort_value->Init(nullptr,  // This OrtValue is "None"
+                           type,
+                           type->GetDeleteFunc());
+  }
+
+  return Status::OK();
+}
+
+Status OptionalHasElement::Compute(OpKernelContext* ctx) const {
+  const auto* input_ort_value = ctx->GetInputOrtValue(0);
+
+  // Output is a scalar
+  auto* output_tensor = ctx->Output(0, {});
+  output_tensor->MutableData<bool>()[0] = input_ort_value->HasElement();
+
+  return Status::OK();
+}
+
+Status OptionalGetElement::Compute(OpKernelContext* ctx) const {
+  const auto* input_ort_value = ctx->GetInputOrtValue(0);
+
+  ORT_ENFORCE(input_ort_value->HasElement(),
+              "Trying to use OptionalGetElement on an optional type OrtValue which is None");
+
```

Comment:
Would returning Status be more appropriate here as I think it's possible for this to occur due to runtime input? It's also more graceful in a minimal build with exceptions disabled (error message vs abort())

Reply:
Fixed